plasmapy/diagnostics/tests/test_charged_particle_radiography.py

```diff
@@ -459,26 +458,203 @@ def test_run_options():
+                },
+            ),
+        ],
+    )
+    def test_intensity_histogram(self, args, kwargs, expected):
+        """Test several valid use cases."""
+        results = cpr.synthetic_radiograph(*args, **kwargs)
+
+        assert len(results) == 3
+
+        x = results[0]
+        assert isinstance(x, u.Quantity)
+        assert x.unit == u.m
+        assert x.shape == (expected["bins"][0],)
+        assert np.isclose(np.min(x), expected["xrange"][0], rtol=1e4)
+        assert np.isclose(np.max(x), expected["xrange"][1], rtol=1e4)
+
+        y = results[1]
+        assert isinstance(y, u.Quantity)
+        assert y.unit == u.m
+        assert y.shape == (expected["bins"][1],)
+        assert np.isclose(np.min(y), expected["yrange"][0], rtol=1e4)
+        assert np.isclose(np.max(y), expected["yrange"][1], rtol=1e4)
+
+        histogram = results[2]
+        assert isinstance(histogram, np.ndarray)
+        assert histogram.shape == expected["bins"]
+
+    def test_optical_density_histogram(self):
+        """
+        Test the optical density calculation is correct and stuffed
+        with numpy.inf when the intensity is zero.
+        """
+        bins = (200, 60)
+        size = np.array([[-1, 1], [-1, 1]]) * 30 * u.cm
+
+        intensity_results = cpr.synthetic_radiograph(
+            self.sim_results, size=size, bins=bins
+        )
+        od_results = cpr.synthetic_radiograph(
+            self.sim_results, size=size, bins=bins, optical_density=True
+        )
+
+        assert np.allclose(intensity_results[0], od_results[0])
+        assert np.allclose(intensity_results[1], od_results[1])
+
+        intensity = intensity_results[2]
+        zero_mask = intensity == 0
+        i0 = np.mean(intensity[~zero_mask])
+        od = -np.log10(intensity / i0)
+
+        assert np.allclose(od[~zero_mask], od_results[2][~zero_mask])
+        assert np.all(np.isposinf(od_results[2][zero_mask]))
+
+
+def test_saving_output(tmp_path):
+
+    path = os.path.join(tmp_path, "temp.npz")
 
-    # CREATE A RADIOGRAPH OBJECT
     grid = _test_grid("electrostatic_gaussian_sphere", num=50)
     source = (0 * u.mm, -10 * u.mm, 0 * u.mm)
     detector = (0 * u.mm, 200 * u.mm, 0 * u.mm)
 
-    sim = prad.SyntheticProtonRadiograph(grid, source, detector, verbose=False)
+    sim = cpr.Tracker(grid, source, detector, verbose=False)
+
+    # Test that output cannot be saved prior to running
+    with pytest.raises(RuntimeError):
+        sim.results_dict
+
     sim.create_particles(1e4, 3 * u.MeV, max_theta=10 * u.deg)
     sim.run(field_weighting="nearest neighbor")
 
-    size = np.array([[-1, 1], [-1, 1]]) * 30 * u.cm
-    bins = [200, 60]
+    res1 = sim.results_dict
+
+    # Save result
+    sim.save_results(path)
+
+    # Load result
+    res2 = dict(np.load(path, "r", allow_pickle=True))
 
-    # Test size is None, default bins
-    h, v, i = sim.synthetic_radiograph()
+    assert np.all(res1["x"] == res2["x"])
```

Comment:
```suggestion
    assert set(results_1.keys()) == set(results_2.keys())
    for key in results_1.keys():
        assert np.allclose(results_1[key], results_2[key])
```

Make tests more robust.  Everything about the dictionary needs to be tested.